### PR TITLE
Make fileystem module idemponent. Compare devicesize and filesystemsize.

### PR DESCRIPTION
**Pull request by alxgu**
#### Issue Type:

Bug PR
#### Ansible Version:

ansible 2.0.0 (stable-2.0.1 2c3814a012) last updated 2016/01/12 16:21:47 (GMT +200)
  lib/ansible/modules/core: (stable-2.0.1 c9735fddd8) last updated 2016/01/12 16:21:36 (GMT +200)
  lib/ansible/modules/extras: (stable-2.0.1 6dee19ee24) last updated 2016/01/12 16:22:05 (GMT +200)
#### Summary:

Today is the changed status alwasy True if you use resizefs=True
server123 | success >> {
    "changed": true,
    "msg": ""
}
#### Expected Results:

dev == fs
server123 | success >> {
    "changed": false,
    "msg": "ext3 filesystem is using the whole device /dev/vg00/oracle.test"
}
dev != fs
server123 | success >> {
    "changed": true,
    "msg": "Filesystem at /dev/vg0/test is mounted on /test; on-line resizing required\nold desc_blocks = 1, new_desc_blocks = 1\nPerforming an on-line resize of /dev/vg0/test to 1048576 (4k) blocks.\nThe filesystem on /dev/vg0/test is now 1048576 blocks long.\n\n"
}

Maybe someone has know how to implement this for btrfs. But right now is the code for btrfs broken, too.

### CODE REVIEW

File: system/filesystem.py [hunk 1]
```diff
@@ -63,6 +63,44 @@
...
+def _get_dev_size(dev, module):
+    """ Return size in bytes of device. Returns int """
+    blockdev_cmd = module.get_bin_path("blockdev", required=True)
+    rc, devsize_in_bytes, err = module.run_command("%s %s %s" % (blockdev_cmd, "--getsize64", dev))
+    return int(devsize_in_bytes)
+
+
+def _get_fs_size(fssize_cmd, dev, module):
+        """ Return size in bytes of filesystem on device. Returns int """
+        fssize_cmd = module.get_bin_path(fssize_cmd, required=True)
```
Comment (abulimov) [inline system/filesystem.py]: If we assign full binary path to a new variable, for example `cmd`, we can later check for exact match against fssize_cmd, like `if 'tune2fs' == fssize_cmd`, and avoid useless `in` check.

File: system/filesystem.py [hunk 2]
```diff
@@ -63,6 +63,44 @@
...
+        if 'tune2fs' in fssize_cmd:
+            # Get Block count and Block size
+            rc, size, err = module.run_command("%s %s %s" % (fssize_cmd, '-l', dev))
```
Comment (abulimov) [inline system/filesystem.py]: Maybe we could check `rc` to avoid trying to parse incorrect output?

File: system/filesystem.py [hunk 3]
```diff
@@ -143,10 +188,18 @@ def main():
...
         module.exit_json(changed=False)
     elif fs == fstype and resizefs == True:
-        cmd = module.get_bin_path(growcmd, required=True)
-        if module.check_mode:
-            module.exit_json(changed=True, msg="May resize filesystem")
-        else:
+        # Get dev and fs size and compare
+        devsize_in_bytes = _get_dev_size(dev, module)
+        fssize_in_bytes = _get_fs_size(fssize_cmd, dev, module)
+        fs_smaller = True if fssize_in_bytes < devsize_in_bytes else False
```
Comment (abulimov) [inline system/filesystem.py]: This does not work with Python 2.4

**Comment by abulimov** [comment]
Thanks for this pull request! I like your idea, and will try to test your changes soon.

**Comment by alxgu** [comment]
Thanks for your feedback. I have updated the PR.

**Comment by gregdek** [comment]
Thanks @alxgu. @abulimov please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

File: system/filesystem.py [hunk 4]
```diff
@@ -63,6 +63,50 @@
...
 '''
 
+def _get_dev_size(dev, module):
+    """ Return size in bytes of device. Returns int """
+    blockdev_cmd = module.get_bin_path("blockdev", required=True)
+    rc, devsize_in_bytes, err = module.run_command("%s %s %s" % (blockdev_cmd, "--getsize64", dev))
+    return int(devsize_in_bytes)
+
+
+def _get_fs_size(fssize_cmd, dev, module):
```
Comment (abulimov) [inline system/filesystem.py:73]: Whole function's first level of indentation is wrong - should be 4 spaces, now 8.

**Comment by abulimov** [comment]
I've tested your changes, and all looks good to me, so this can be merged after small style fix. By the way, while testing your code I've found that resizing btrfs never worked - `grow_flag` isn't even used in code. I'll try to fix it myself in a separate pull request.

**Comment by alxgu** [comment]
fixed

**Comment by abulimov** [comment]
shipit

**Comment by gregdek** [comment]
Thanks again to @alxgu for this PR, and thanks @abulimov for reviewing. Marking for inclusion.

**Pull request merged**